### ISSUE TITLE
error when i try to convert facenet.pb to facenet.onnx using tf2onnx.convert

### ISSUE THREAD

**Issue by awaradh** [onnx/tensorflow-onnx#1207]
i'm trying to convert tensorflow model to onnex model, but i get this error 
RuntimeError: Shape inference error(s): (op_type:Cast, node name: InceptionResnetV1/Bottleneck/BatchNorm/Reshape_1__503): [TypeInferenceError] Value of attribute to not specified
any help please
thanks

**Comment by TomWildenhain-Microsoft** [comment]
Seems like a "Cast" node is being created without the "to" attribute.  Can you upload the model you are trying to convert?

**Comment by guschmue** [comment]
There is no enough info here to look at this. Do you have at least the stack trace from the error?

**Comment by annabeth97c** [comment]
Hi! Not the original poster, but I'm facing a similar issue.

I am trying to convert the ICNet tensorflow model (from [this](https://github.com/hellochick/ICNet-tensorflow) repo) to a trt engine. I saved the model first, then tried using tf2onnx to convert it to an onnx model. When I do not fill in the opset argument, I get a value error 

> ValueError: get tensor value: pad_to_bounding_box/Reshape must be Const

I found some other people faced this issue and setting opset to 11 helped them so I followed that. A runtime error is thrown which is very similar to the one mentioned by @awaradh above.

But the process continues and generates an onnx. I then tried to convert this onnx model to a tensorrt engine but it fails to be parsed by the tensorrt onnx parser. So I think it might be this runtime error that is giving me a flawed onnx model

Here is the error I get:

> Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tf2onnx/schemas.py", line 157, in infer_onnx_shape_dtype
    inferred_model = shape_inference.infer_shapes(model_proto)
  File "/usr/local/lib/python3.5/dist-packages/onnx/shape_inference.py", line 35, in infer_shapes
    inferred_model_str = C.infer_shapes(model_str, check_type)
RuntimeError: Shape inference error(s): (op_type:Cast, node name: pad_to_bounding_box/Reshape__26): [TypeInferenceError] Value of attribute to not specified

**Comment by TomWildenhain-Microsoft** [comment]
@annabeth97c can you try running the converted model in onnx runtime?  That will indicate whether the error is with tf2onnx or tensorRT.

**Comment by annabeth97c** [comment]
@TomWildenhain-Microsoft thanks for the prompt response! 

When I try to load the model using an onnxruntime InferenceSession I get this error:

> Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/onnxruntime/capi/session.py", line 158, in __init__
    self._load_model(providers or [])
  File "/usr/local/lib/python3.5/dist-packages/onnxruntime/capi/session.py", line 177, in _load_model
    self._sess.load_model(providers)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. Error in Node:strided_slice_6 : No Op registered for Slice with domain_version of 13

**Comment by annabeth97c** [comment]
Sorry I realised I was using the wrong onnx file. The one which gave me the above error was one generated with opset 13. 

This is the error I get with the onnx model generated with opset 11

> Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/onnxruntime/capi/session.py", line 158, in __init__
    self._load_model(providers or [])
  File "/usr/local/lib/python3.5/dist-packages/onnxruntime/capi/session.py", line 177, in _load_model
    self._sess.load_model(providers)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Node (conv5_3_sum) Op (Sum) [ShapeInferenceError] Incompatible dimensions

**Comment by TomWildenhain-Microsoft** [comment]
Interesting.  The first error might be a problem with ORT since there definitely is a Slice op in opset 13.  The second looks like a converter error.  Can you please upload the saved model you are converting as a zip along with the two onnx models you got?

**Comment by annabeth97c** [comment]
Sorry for the delay! Here you go! 

Its a google google drive link as the file is too big for git! 

https://drive.google.com/file/d/1WEPTgct971J5zIme0PhoH-VI4Wl7iPmk/view?usp=sharing

**Comment by annabeth97c** [comment]
Also forgot to mention, the tags used were ['train', 'serve']

**Comment by TomWildenhain-Microsoft** [comment]
I'm able to reproduce your error for opset 11.  For opset 13 I get a different error:
`Error in Node:ExpandDims : Node (ExpandDims) has input size 1 not in range [min=2, max=2]`
I believe you are using an old version of onnxruntime which does not support opset 13.  Regardless, there is a problem with both models.  

Please run `pip install --upgrade onnxruntime`.  Try converting using the latest tf2onnx from git:
`pip uninstall tf2onnx`
`pip install git+https://github.com/onnx/tensorflow-onnx`

Also please upload the saved model you are trying to convert.

**Comment by annabeth97c** [comment]
Hi! The saved model is in the same zip, in the directory '1'

**Comment by TomWildenhain-Microsoft** [comment]
Oh oops, didn't see it.  Just tried with the latest version from master and the new model seems to work.  Can you try again using tf2onnx from master?  The problem might be caused by an issue that was recently fixed.

`pip uninstall tf2onnx`
`pip install git+https://github.com/onnx/tensorflow-onnx`

However my model looks very different from yours.  Are you sure you uploaded the correct model?  Yours looks like:
![image](https://user-images.githubusercontent.com/67606533/107318007-bc686280-6a69-11eb-8183-329fc25d3963.png)

However my opset13 looks like:
![image](https://user-images.githubusercontent.com/67606533/107318066-e0c43f00-6a69-11eb-9706-6aff47d8b5e3.png)

**Comment by TomWildenhain-Microsoft** [comment]
Also are you using the python API or command line interface?  What code/command are you using to convert?

**Comment by annabeth97c** [comment]
Hi! Yep I managed to convert using the latest version from master! I am facing issues with converting the onnx to tensorrt because of the node "Sub" after the Placeholder in your first pic! The new onnx I generated still looks like that with the Placeholder and Sub node.

 Can I ask, what do you mean by your opset13? Its your onnx after converting the saved model?

I am using this command to convert 

> python -m tf2onnx.convert --saved-model ./2 --opset 11 --output model.onnx

I saved the model with only the serve tag in case there were issues because of that

**Comment by TomWildenhain-Microsoft** [comment]
Yes by "my opset13" I meant the model I get when I convert your saved_model using opset 13.  I don't know why it is so different from your model.  Maybe our tensorflow versions are different and mine is applying more aggressive optimizations (removing the subtract 0, etc).  I'm on TF 2.4.1, which TF are you using?  Can you try updating the version of TF for the converter?

Yeah I also had issues with the multiple tags that I had to hack around.

What tensorRT error are you getting?  Can you confirm that the model works in onnxruntime?  TensorRT only has partial support for onnx and can have issues with valid models (normally if the shapes are non-constant).

**Comment by annabeth97c** [comment]
Hi! So sorry for the late reply!

Yes the model worked in onnxruntime.

The tensorrt error looked like this

>  While parsing node number 0 [Pad -> "pad:0"]:ERROR:/workspace/repos/onnx-tensorrt/ModelImporter.cpp:537
In function importModel:[5] Assertion failed: tensors.count(input_name)

To try and replicate your result, I generated the ONNX using tensorflow 2.3.0 and it seems to have worked correctly. My model looks like the picture of your opset 13 conversion!

I was still getting errors at the first padding layer when trying to convert the onnx to tensorrt (The input shape was fixed). I found that the version of tensorrt I was using (v5.1.5) had given others issues like this previously and that these issues have been patched in future versions so I used tensorrt v7.0 and was able to convert the model successfully using opset 9 (Pad layers in opset 11 models give errors)

Thank you so much for your help!
